server/sonar-server/src/main/java/org/sonar/server/computation/step/ComputePluginMeasuresStep.java

```diff
@@ -0,0 +1,197 @@
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.step;
+
+import com.google.common.base.Optional;
+import java.util.List;
+import org.sonar.api.compute.Component;
+import org.sonar.api.compute.Measure;
+import org.sonar.api.compute.MetricsDefinition;
+import org.sonar.api.config.Settings;
+import org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor;
+import org.sonar.server.computation.component.ProjectSettingsRepository;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.measure.MeasureRepository;
+import org.sonar.server.computation.metric.MetricRepository;
+
+import static org.sonar.server.computation.component.Component.Type.FILE;
+import static org.sonar.server.computation.component.ComponentVisitor.Order.PRE_ORDER;
+import static org.sonar.server.computation.measure.Measure.newMeasureBuilder;
+
+/**
+ * TODO
+ */
+public class ComputePluginMeasuresStep implements ComputationStep {
+
+  private final TreeRootHolder treeRootHolder;
+  private final MetricRepository metricRepository;
+  private final MeasureRepository measureRepository;
+
+  // TODO We should not inject MetricsDefinition[] but a list of Metrics (took from the parent pico)
+  private final MetricsDefinition[] metricsDefinitions;
+  private final ProjectSettingsRepository settings;
+
+  public ComputePluginMeasuresStep(TreeRootHolder treeRootHolder, MetricRepository metricRepository, MeasureRepository measureRepository, MetricsDefinition[] metricsDefinitions,
+    ProjectSettingsRepository settings) {
+    this.treeRootHolder = treeRootHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+    this.metricsDefinitions = metricsDefinitions;
+    this.settings = settings;
+  }
+
+  @Override
+  public void execute() {
+    new NewMetricDefinitionsVisitor().visit(treeRootHolder.getRoot());
+  }
+
+  private class NewMetricDefinitionsVisitor extends DepthTraversalTypeAwareVisitor {
+    public NewMetricDefinitionsVisitor() {
+      super(FILE, PRE_ORDER);
+    }
+
+    @Override
+    public void visitAny(org.sonar.server.computation.component.Component component) {
+      MetricsDefinition.NewMetricContext newMetricContext = new MetricsDefinition.NewMetricContext();
+      for (MetricsDefinition metricsDefinition : metricsDefinitions) {
+        metricsDefinition.define(newMetricContext);
+      }
+
+      // TODO use DirectAcyclicGraph to execute measure computation in the right order
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)